scipy/linalg/_decomp_update.pyx

```diff
@@ -0,0 +1,2416 @@
+    cdef int p1 = p
+    cdef int p_eco, p_full
+    cdef int typecode, m, n, info
+    cdef int qs[2]
+    cdef int rs[2]
+    cdef bint economic, qisF = False 
+
+    if which == 'row':
+        q1, r1, typecode, m, n, economic = validate_qr(Q, R, overwrite_qr,
+                NPY_ANYORDER, overwrite_qr, NPY_ANYORDER, check_finite)
+        if not (-m <= k1 < m):
+            raise ValueError('k is not a valid index')
+        if k1 < 0:
+            k1 += m
+        if k1 + p1 > m or p1 <= 0: 
+            raise ValueError('p out of range')
+        if economic:
+            if not cnp.PyArray_ISONESEGMENT(q1):
+                q1 = PyArray_FromArraySafe(q1, NULL, cnp.NPY_F_CONTIGUOUS)
+                qisF = True
+            elif cnp.PyArray_CHKFLAGS(q1, cnp.NPY_F_CONTIGUOUS):
+                qisF = True
+            else:
+                qisF = False
+            if m-p >= n:
+                p_eco = p1
+                p_full = 0
+            else:
+                p_eco = m-n
+                p_full = p1 - p_eco
+            if typecode == cnp.NPY_FLOAT:
+                info = thin_qr_row_delete(m, n, <float*>extract(q1, qs), qs, qisF,
+                    <float*>extract(r1, rs), rs, k1, p_eco, p_full)
+            elif typecode == cnp.NPY_DOUBLE:
+                info = thin_qr_row_delete(m, n, <double*>extract(q1, qs), qs, qisF,
+                    <double*>extract(r1, rs), rs, k1, p_eco, p_full)
+            elif typecode == cnp.NPY_CFLOAT:
+                info = thin_qr_row_delete(m, n, <float_complex*>extract(q1, qs), qs, qisF,
+                    <float_complex*>extract(r1, rs), rs, k1, p_eco, p_full)
+            else:  #  cnp.NPY_CDOUBLE
+                info = thin_qr_row_delete(m, n, <double_complex*>extract(q1, qs), qs, qisF,
+                    <double_complex*>extract(r1, rs), rs, k1, p_eco, p_full)
+            if info == 1:
+                return q1[p_full:-p_eco, p_full:], r1[p_full:,:]
+            elif info == MEMORY_ERROR:
+                raise MemoryError('malloc failed')
+            else:
+                raise ValueError('Reorthogonalization Failed, unable to perform row deletion.')
+        else:
+            if typecode == cnp.NPY_FLOAT:
+                qr_block_row_delete(m, n, <float*>extract(q1, qs), qs,
+                    <float*>extract(r1, rs), rs, k1, p1)
+            elif typecode == cnp.NPY_DOUBLE:
+                qr_block_row_delete(m, n, <double*>extract(q1, qs), qs,
+                    <double*>extract(r1, rs), rs, k1, p1)
+            elif typecode == cnp.NPY_CFLOAT:
+                qr_block_row_delete(m, n, <float_complex*>extract(q1, qs), qs,
+                    <float_complex*>extract(r1, rs), rs, k1, p1)
+            else:  # cnp.NPY_CDOUBLE:
+                qr_block_row_delete(m, n, <double_complex*>extract(q1, qs), qs,
+                    <double_complex*>extract(r1, rs), rs, k1, p1)
+            return q1[p1:, p1:], r1[p1:, :]
+    elif which == 'col':
+        if p1 > 1:
```

Comment:
Yes, that correct.  There isn't a way to remove the rows in one pass, as there is for the columns, so the block row code really just loops over the rows to be removed.  I can add a note.